### ISSUE TITLE
Special Characters in Filenames

### ISSUE THREAD

**Issue by andre111** [caprica/vlcj#415]
I cann't seem to get files with special characters in their name to work.

These are the error messages I recieved:

```
[000000005d3ccf10] filesystem access error: cannot open file Y:\Disney Animationsfilme\Aladin und der K
[000000005d3ccf10] core access error: Einlesen der Datei fehlgeschlagen
[000000005d3ccf10] core access error: VLC konnte die Datei 
[00000000002c5bf0] core input error: open of `file:///Y:/Disney%20Animationsfilme/Aladin%20und%20der%20K%F6nig%20der%20Diebe.mkv' failed
[00000000002c5bf0] core input error: Ihre Eingabe konnte nicht geöffnet werden
[00000000002c5bf0] core input error: VLC kann die Medienadresse 'file:///Y:/Disney%20Animationsfilme/Aladin%20und%20der%20K%F6nig%20der%20Diebe.mkv' nicht öffnen. Für Details bitte im Fehlerprotokoll nachsehen.
```

Searching with Google lead me to this: http://stackoverflow.com/questions/11507633/vlcj-mlr-special-character where you said it should simply work on vlcj 2.4.0 and up. I'm currently using 3.0.1 and it wont work no matter what I try.

**Comment by caprica** [comment]
Please post here exactly the filename as you typed it in your Java program.

**Comment by caprica** [comment]
Similar to #173.

**Comment by andre111** [comment]
The path gets loaded from a xml file using jdom 2.0.5

System.out.println prints exactly the same as the Path in the xml file:
Y:\Disney Animationsfilme\Aladin und der König der Diebe.mkv

**Comment by caprica** [comment]
The vlcj log would be really helpful.

I know that filename will work on Linux. Windows seems much more problematic for this stuff.

**Comment by andre111** [comment]
How do I enable it/where do I find it?

**Comment by andre111** [comment]
Nevermind. The code from the linked issue is working perfectly fine.
Using this now:
```
File file = new File(url);
URI uri = file.toURI();
String asciiURI = uri.toASCIIString();
if(Platform.isWindows()) {
	asciiURI = asciiURI.replace("file:/", "file:///");
}
```

**Comment by caprica** [comment]
http://capricasoftware.co.uk/#/projects/vlcj/tutorial/logging

**Comment by andre111** [comment]
Should I post a log anyway? If so it has to wait until tomorrow. It's really late for me.

**Comment by caprica** [comment]
Whenever you want, it might help.

I can't immediately see the difference between your fix from the linked issue and this code in vlcj:
```
result = new File(mrl).toURI().toASCIIString().replaceFirst("file:/", "file:///");
```
But then again, it's very late for me too...

**Comment by andre111** [comment]
Yeah, I just noticed that to. My fix can even be compacted to the very same line. Still it produces different results as I allready pointed out. I don't even know whats going on.

**Comment by andre111** [comment]
There seems to be something wrong with the checks before converting the Path in MediaResourceLocator.encodeMrl.
This:
```
String asciiURI = new File(url).toURI().toASCIIString().replaceFirst("file:/", "file:///");
System.out.println(asciiURI);
System.out.println(MediaResourceLocator.encodeMrl(url));
```
Prints that:
```
file:///Y:/Disney%20Animationsfilme/Aladin%20und%20der%20K%C3%B6nig%20der%20Diebe.mkv
Y:\Disney Animationsfilme\Aladin und der König der Diebe.mkv
```

Just as a remainder. I'm using vlcj-3.0.1 but it doesn't seem like you made any important changes on that part of the code anyway.

**Comment by caprica** [comment]
Yeah, this is a bug really.

The problem is coming up with a reliable heuristic that handles non-file URL cases as well as this case in a robust manner.

Aside from that, as far as I know backslashes are always invalid in URIs no matter what platform.

**Comment by caprica** [comment]
Do you have something that works without the file:// prefix?

**Comment by andre111** [comment]
Where do you mean? In mediaPlayer.playMedia or in the URI?

Maybe just using file.exists() for detecting whether a URL is a (local) file or something else could work.

**Comment by caprica** [comment]
I have been reviewing the current implementation and I don't know what I was thinking. It needs to be revisited.

In your comment above where you posted something that works from you and you referenced the other linked issue, I was wondering if you tried removing the file:// prefix entirely and seeing if it still worked.

**Comment by caprica** [comment]
Please try this...

Add Guava to your project:

```
<dependency>
    <groupId>com.google.guava</groupId>
    <artifactId>guava</artifactId>
    <version>18.0</version>
</dependency>        
```

Then use this:

```
String encodedMrl = new PercentEscaper(":_.\\", false).escape("Y:\Disney Animationsfilme\Aladin und der König der Diebe.mkv");
mediaPlayer.playMedia(encodedMrl);
```

I don't really want to write my own encoder, so I'd prefer to use something like this. If this works, I can either add Javadoc to vlcj, or I could probably try and take only the relevant classes from Guava (Apache 2.0 licensed) and include them in vlcj (Guava is a 2.4Mb jar file which is a bit large).

**Comment by andre111** [comment]
Doesn't seem to work.
It results in this mrl:
Y:\Disney Animationsfilme\Aladin und der K%C3%B6nig der Diebe.mkv
Which results in the same error from VLC.

Adding "file:///" at the start of this mrl fixes it again. But then you would still need to check whether it actually is a file mrl or something elso

**Comment by caprica** [comment]
OK, thanks for your testing and research. I am still trying to find something simple and robust.

**Comment by caprica** [comment]
If you go to a command-line, and type:
```
vlc "Y:\Disney Animationsfilme\Aladin und der König der Diebe.mkv"
```
Does that work?

**Comment by andre111** [comment]
Yes typing it into the windows command prompt (cmd.exe) works perfectly fine

**Comment by caprica** [comment]
Thanks, this seems like a Java-to-native issue. I'll investigate some more...

**Comment by caprica** [comment]
I am *very* confused about this issue.

I have an en-gb Locale WindowsXP (32-bit) running in a virtual machine.
I use VLC 2.2.1.

I have a filename: `c:\\König.avi`

I also modified vlcj so that it uses the supplied MRL directly, and *does not convert it at all*.

It plays just fine!

Here's part of the native log:

```
INPUT MRL IS c:\König.avi
[031786c4] core input debug: Creating an input for 'K�nig.avi'
[031786c4] core input debug: `file:///c:/K%C3%B6nig.avi' gives access `file' demux `' path `/c:/K%C3%B6nig.avi'
[031786c4] core input debug: specified demux `any'
[031786c4] core input debug: creating demux: access='file' demux='any' location='/c:/K%C3%B6nig.avi' file='c:\K�nig.avi'
[031e8d54] core demux debug: looking for access_demux module matching "file": 12 candidates
[031e8d54] core demux debug: no access_demux modules matched
[031786c4] core input debug: creating access 'file' location='/c:/K%C3%B6nig.avi', path='c:\K�nig.avi'
[031f4f04] core access debug: looking for access module matching "file": 21 candidates
[031f4f04] filesystem access debug: opening file `c:\K�nig.avi'
[031f4f04] core access debug: using access module "filesystem"
[031e6934] core stream debug: Using stream method for AStream*
[031e6934] core stream debug: starting pre-buffering
[031e6934] core stream debug: received first data after 0 ms
```
The strange character in the log indeed appears as some unprintable character in my Eclipse console window, but the video does start playing.

So according to my test, no conversion at all is necessary and the vlcj call to convert the MRL should be removed!

**Comment by caprica** [comment]
If you have the vlcj sources in your project, please go to SimpleMedia.java and change it:

```
    public SimpleMedia(String mrl, String... mediaOptions) {
        super(mediaOptions);
        // Encode the MRL if necessary (e.g. if it is a local file that contains Unicode characters)
        this.mrl = mrl; // MediaResourceLocator.encodeMrl(mrl);
    }
```

**Comment by caprica** [comment]
(Note to self: In future, SimpleMedia should not encode anyway, and if encoding eventually is still needed it should be moved to its own Media sub-class.)

**Comment by andre111** [comment]
The problem is: Due to the exception thrown in MediaResourceLocator.encodeMrl method it allready doesn't modify the MRL. But it still doesn't work like that for me. I have to do the encoding which would be done in MediaResourceLocator.encodeMrl myself to make it work. 

(Because I don't need to check for URI other than file locations I just copied "String asciiURI = new File(url).toURI().toASCIIString().replaceFirst("file:/", "file:///");" without the prior checks.)

It definitelly needs encoding to work for me(Windows 7(64 bit) - German Locale). From the Log you posted it seems like VLC does the MRL encoding itself but I noticed one difference:

Encoding the MRL in Java(or letting VLC do it in the virtual Windows XP) "König" results in:
K%C3%B6nig (which is the Unicode UTF8 representation and works)
While for me the Log from vlc doing the encoding itself prints this instead:
K%F6nig (which uses HTML encoding which doesn't seem to work)

**Comment by caprica** [comment]
I understand the part about the exception, it's clearly a bug in the current implementation.

I just think there's something else going on here.

Do you see the same problem if you use the filename directly, rather than using JDOM (as you originally posted)?

**Comment by andre111** [comment]
Directly writing the Filepath as a string in a .java encoded with UTF-8 results in the same error.

**Comment by caprica** [comment]
If you go to a terminal and type "chcp", what does it say?

**Comment by andre111** [comment]
850.
Sorry, don't have time for more detailed answers right now.

**Comment by caprica** [comment]
That's fine, thanks.

I am trying hard to reproduce this bug so I can fix it properly and robustly. So far I have no way to reproduce it on the Windows platforms available to me.

**Comment by HagiHagi** [comment]
I don't know if the above discussion is closed or not.
There is a difference between starting the project inside of netbeans and starting the jar file!

I'm using netbeans 8.1. and WIndows 10 64bit.
The project's encoding is UTF-8

I use following code:
			URI uri = m_MediaSource.getFile().toURI();
			stringURI = uri.toASCIIString();
			stringURI = stringURI.replace("file:/", "file:///");
System.out.println("prepare Media " + stringURI);
			boolean boolSuccess = m_MediaPlayer.prepareMedia(stringURI);
			if (boolSuccess) {
System.out.println("parse Media ");
				m_MediaPlayer.parseMedia();
System.out.println("start Media ");
				boolSuccess = m_MediaPlayer.start();
			}
System.out.println("MediaPlayer started ");

For some mediaplayer events I added a System.out

1.) starting the project with netbeans all works fine
2.) starting the jar file an error occurs

add 1.) Output starting with netbeans
prepare Media file:///F:/04_Laser%20Font%C3%A4ne.mp4
parse Media 
LiveShow_MediaPlayer_Video_VLC mediaChanged file:///F:/04_Laser%20Font%C3%A4ne.mp4
LiveShow_MediaPlayer_Video_VLC newMedia 
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 13
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
start Media 
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC opening 
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 13
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
MediaPlayer started 

add 2.) Output starting the jar file
prepare Media file:///F:/04_Laser%20Font%C3%A4ne.mp4
parse Media
LiveShow_MediaPlayer_Video_VLC mediaChanged file:///F:/04_Laser%20Font%C3%A4ne.mp4
LiveShow_MediaPlayer_Video_VLC newMedia
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 13
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
start Media
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC opening
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 12
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 13
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
LiveShow_MediaPlayer_Video_VLC mediaMetaChanged 15
MediaPlayer started
[0000000013063040] core vout display error: Failed to change zoom
[0000000013063040] core vout display error: Failed to set on top
[0000000013063040] core vout display error: Failed to change source AR
[000000002302c340] filesystem access error: cannot open file F:\04_Laser Font�ne.mp4 (No such file or directory)
[000000002302c340] core access error: Einlesen der Datei fehlgeschlagen
[000000002302c340] core access error: VLC konnte die Datei „F:\04_Laser Font�ne.mp4“ nicht öffnen (No such file or directory).
[00000000130ac7a0] core input error: open of `file:///F:\04_Laser Font�ne.mp4' failed
[00000000130ac7a0] core input error: Ihre Eingabe konnte nicht geöffnet werden
[00000000130ac7a0] core input error: VLC kann die Medienadresse 'file:///F:\04_Laser Font�ne.mp4' nicht öffnen. Für Details bitte im Fehlerprotokoll nachsehen.

Perhaps it helps.

**Comment by caprica** [comment]
It's open until it's fixed properly!

Thanks for the feedback.

Is it possible to see 1) the exact command-line that is used to start Netbeans; and 2) the exact command-line that is used to start your application inside Netbeans. This should include in particular system properties, environment variables and so on.

**Comment by HagiHagi** [comment]
Thanks for the fast answer.
I think it is not a problem with command-line or other vm options.

I tried several things, at least I checked the encoding of the jvm:

System.out.println("Encoding " + Charset.defaultCharset().name());

With following results:
a) Start the project under Windows using the 'start project' inside netbeans: Encoding UTF-8
b) Start the project under Windows using the terminal 'java -jar ...' : Encoding windows-1252
c) Start the project under MAC using the terminal 'java -jar ...' : Encoding UTF-8

The jvm's encoding may be different under different conditions. This will affect calls like myString.getBytes(). It is always better to use myString.getBytes(<charset name>);

I don't have a look to the vlcj source code, maybe there exists function-calls with dependencies to the defaultCharset.

**Comment by caprica** [comment]
So the only think that works consistently is when using UTF-8?

**Comment by caprica** [comment]
According to https://jna.java.net/javadoc/overview-summary.html, JNA will use the platform default encoding when passing Java Strings as native strings.

This can be controlled using the "jna.encoding" system property, the example given suggests to set it to "UTF8".

It would be very interesting to know if this solves all the problems described in this issue...

**Comment by HagiHagi** [comment]
Ok we have a working solution!
Setting the jna.encoding to UTF-8 will solve one part of the problems.

The other part is to replace the backslashes with normal slashes. Because in the MediaResourceLocator.encodeMrl(String mrl) you use 
Uri uri = new URI(mrl);
This will throw following exception, when mrl contains a backslash \ :
 (java.net.URISyntaxException) java.net.URISyntaxException: Illegal character in path at index 10: file:///F:\Fontäne.mp4 
This is caused by a violation of [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt)
-> you should first replace all '\' with '/' 

**Here is the working solution**
Source code:
//a) Add to your main function
System.setProperty("jna.encoding", "UTF8");

//in the code use
...
File o_File = new File("F:\Fontäne.mp4");
String stringURI = "file:///" + o_File().getAbsolutePath();
stringURI = stringURI.replace("\\", "/");
myDirectMediaPlayer.prepareMedia(stringURI );
...

**Comment by HagiHagi** [comment]
In the above code the editor swallows one \, of course you should use:
stringURI = stringURI.replace("\\\", "/");

**Comment by caprica** [comment]
I'm not 100% sure encodeMrl still needs to get the URI, if the encoding is right.

**Comment by HagiHagi** [comment]
Shure - the function will be called, when you use my excample code.
On my system The stringURI is a UTF-8 string -> the new URI(mrl) will be called.
It makes only trouble on windows when the file's path is used, because this contains the backslash.

I think encodeMrl is not necessary if the uri.getScheme is ok! But to check the scheme, the URI must be instanciated.
I checked several situations all works fine, even if mrl parameter is encoded as unicode or ASCII.

Perhaps it is better to add a function prepareMedia(File p_File) to the MediaPlayer's code, than you have a better control - how to create an URI string for files.

**Comment by caprica** [comment]
play/prepareMedia can now take a "Media" object, so I can simply add a new sub-class that accepts a File and as you suggest and encapsulate all this. There's no need for new API functions, which is good.

I'll look into this at the weekend to try and get a 100% correct fix.

Thanks for the testing and investigations you did on this.

**Comment by caprica** [comment]
Closing this since #645 is more current and is likely the same problem. Should be fixed in next version of vlcj, likely 3.11.0.
